### PR TITLE
Coverity fixes for 2021-05-17

### PR DESCRIPTION
**Pull request by paulidale**
Fixes for coverity issues: 1484883, 1484884, 1484885, 1484886 and 1484888.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/p_seal.c [hunk 1]
```diff
@@ -34,15 +34,19 @@ int EVP_SealInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *type,
     if (EVP_CIPHER_CTX_rand_key(ctx, key) <= 0)
         return 0;
 
-    if (EVP_CIPHER_CTX_iv_length(ctx)
-            && RAND_bytes(iv, EVP_CIPHER_CTX_iv_length(ctx)) <= 0)
+    len = EVP_CIPHER_CTX_iv_length(ctx);
```
Comment (slontis) [inline crypto/evp/p_seal.c:37]: Arrg - this is a very libctx unfriendly function...

File: crypto/evp/p_seal.c [hunk 2]
```diff
@@ -34,15 +34,19 @@ int EVP_SealInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *type,
...
+    if (len < 0)
         goto err;
 
     if (!EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv))
         goto err;
 
     for (i = 0; i < npubk; i++) {
-        size_t keylen = EVP_CIPHER_CTX_key_length(ctx);
+        size_t keylen = len;
         EVP_PKEY_CTX *pctx = NULL;
```
Comment (slontis) [inline crypto/evp/p_seal.c:50]: This cant be declared here.. There are goto paths below that will cause mem leaks.

File: crypto/provider_core.c [hunk 3]
```diff
@@ -315,6 +314,7 @@ static OSSL_PROVIDER *provider_new(const char *name,
         return NULL;
     }
 
+    prov->refcnt = 1; /* 1 One reference to be returned */
```
Comment (slontis) [inline crypto/provider_core.c:317]: This seems to be the pattern we use elsewhere..

File: crypto/evp/p_seal.c (hunk 1 above)
Comment (paulidale) [reply crypto/evp/p_seal.c:37]: Very.  Fix coming.

**Comment by paulidale** [comment]
Added a commit that makes EVP_SealInit library context aware and avoids the memory leak on error.

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -860,6 +868,8 @@ static int test_EVP_Enveloped(void)
 
     ret = 1;
 err:
+    if (n != 0)
```
Comment (slontis) [inline test/evp_extra_test.c:871]: This if check is no longer required?

Comment (paulidale) [reply test/evp_extra_test.c:871]: Safer to include it :)

**Review by slontis — APPROVED** [review]
LGTM

**Comment by slontis** [comment]
@DDvO - FYI..

**Comment by DDvO** [comment]
> @DDvO - FYI..

Thanks @slontis for letting me know and cleaning up the code structure.
CID 1484883 was a false positive in the sense that `if (value != NULL ...)` was not wrong but needless and thus misleading.
I like your re-structuring of `http_server_get_asn1_req()` because the code is now more linear and simpler to read.
You could have mentioned as part of the extended message of the commit that this 'fixes' CID 1484883.

**Comment by paulidale** [comment]
I marked 1484883 as a false positive in coverity and didn't include the CID in the PR.  I've amended the description to include this.

**Comment by DDvO** [comment]
> I marked 1484883 as a false positive in coverity and didn't include the CID in the PR. I've amended the description to include this.

Thanks Pauli.
I just realized that it was you (not Shane) who is doing this PR.

**Comment by paulidale** [comment]
No problem, I figure it was an innocent mistake :)

File: apps/lib/http_server.c [hunk 5]
```diff
@@ -433,36 +433,33 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
-            (void)http_server_send_status(cbio, 400, "Bad Request");
-            goto out;
         }
-
+        *line_end = '\0';
+        /* https://tools.ietf.org/html/rfc7230#section-6.3 Persistence */
+        if (found_keep_alive != NULL && strcasecmp(key, "Connection") == 0) {
+            if (strcasecmp(value, "keep-alive") == 0)
+                *found_keep_alive = 1;
+            if (strcasecmp(value, "close") == 0)
```
Comment (t8m) [inline apps/lib/http_server.c:460]: could be `else if`

**Review by t8m — APPROVED** [review]

File: apps/lib/http_server.c (hunk 5 above)
Comment (paulidale) [reply apps/lib/http_server.c:460]: Missed this improvement.

Comment (paulidale) [reply apps/lib/http_server.c]: Fixed this too.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master.  Thanks for the reviews and suggestions.

**Pull request closed without merging**